File: skinx/crypto.py
"""Low-level AES-256-CFB8 encryption and Minecraft Bedrock file format."""

import struct
from pathlib import Path

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms
from cryptography.hazmat.decrepit.ciphers.modes import CFB8
from cryptography.hazmat.backends import default_backend

SKINPACK_KEY = b"s5s5ejuDru4uchuF2drUFuthaspAbepE"
SKINPACK_IV = SKINPACK_KEY[:16]
MAGIC = 0x9BCFB9FC  # File magic at bytes 4-7 of header
HEADER_SIZE = 0x100  # 256-byte header before ciphertext
DONT_ENCRYPT = {"manifest.json", "contents.json", "texts", "pack_icon.png"}


def cfb8_encrypt(key: bytes, iv: bytes, data: bytes) -> bytes:
    cipher = Cipher(algorithms.AES(key), CFB8(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    return encryptor.update(data) + encryptor.finalize()


def cfb8_decrypt(key: bytes, iv: bytes, data: bytes) -> bytes:
    cipher = Cipher(algorithms.AES(key), CFB8(iv), backend=default_backend())
    decryptor = cipher.decryptor()
    return decryptor.update(data) + decryptor.finalize()


def read_encrypted_header(filepath: Path):
    """Read the 256-byte header. Returns (uuid, ciphertext)."""
    data = filepath.read_bytes()
    uuid_len = data[16]
    uuid = data[17:17 + uuid_len].decode("utf-8", errors="replace")
    ciphertext = data[HEADER_SIZE:]
    return uuid, ciphertext


def write_encrypted_header(filepath: Path, uuid: str, plaintext: bytes, key: bytes):
    """Write a file with the 256-byte header format."""
    data = write_encrypted_header_bytes(uuid, plaintext, key)
    with open(filepath, "wb") as f:
        f.write(data)


def write_encrypted_header_bytes(uuid: str, plaintext: bytes, key: bytes) -> bytes:
    """Return the 256-byte header format as bytes (no file I/O)."""
    iv = key[:16]
    uuid_bytes = uuid.encode("utf-8")
    ciphertext = cfb8_encrypt(key, iv, plaintext)

    header = bytearray()
    header += struct.pack("<IIQ", 0, MAGIC, 0)
    header += bytes([len(uuid_bytes)])
    header += uuid_bytes
    if len(header) < HEADER_SIZE:
        header += b"\x00" * (HEADER_SIZE - len(header))
    header += ciphertext
    return bytes(header)


def cfb8_decrypt_file(filepath: Path, file_key: bytes = None, data: bytes = None) -> bytes:
    """Decrypt a file, auto-detecting the header format."""
    if data is None:
        data = filepath.read_bytes()
    if len(data) >= 8 and int.from_bytes(data[4:8], "little") == MAGIC:
        ciphertext = data[HEADER_SIZE:]
        key = file_key or SKINPACK_KEY
        return cfb8_decrypt(key, key[:16], ciphertext)
    if file_key:
        return cfb8_decrypt(file_key, file_key[:16], data)
    raise ValueError("No header and no key provided")

File: internal/bedrock/crypto.go
package bedrock

import (
	"crypto/aes"
	"encoding/binary"
	"fmt"
	"os"
)

var SkinpackKey = []byte("s5s5ejuDru4uchuF2drUFuthaspAbepE")
var SkinpackIV = SkinpackKey[:16]

const Magic = 0x9BCFB9FC
const HeaderSize = 0x100

func cfb8Process(key, iv, data []byte, decrypt bool) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	blockSize := block.BlockSize()
	out := make([]byte, len(data))
	t := make([]byte, blockSize)
	copy(t, iv)
	buf := make([]byte, blockSize)
	for i, b := range data {
		block.Encrypt(buf, t)
		out[i] = b ^ buf[0]
		if decrypt {
			t = append(t[1:], b)
		} else {
			t = append(t[1:], out[i])
		}
	}
	return out, nil
}

func CFB8Encrypt(key, iv, data []byte) ([]byte, error) {
	return cfb8Process(key, iv, data, false)
}

func CFB8Decrypt(key, iv, data []byte) ([]byte, error) {
	return cfb8Process(key, iv, data, true)
}

func ReadEncryptedHeader(path string) (uuid string, ciphertext []byte, err error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", nil, err
	}
	if len(data) < HeaderSize {
		return "", nil, fmt.Errorf("file too small for header: %d bytes", len(data))
	}
	uuidLen := int(data[16])
	if 17+uuidLen > len(data) {
		return "", nil, fmt.Errorf("invalid uuid length: %d", uuidLen)
	}
	uuid = string(data[17 : 17+uuidLen])
	ciphertext = data[HeaderSize:]
	return uuid, ciphertext, nil
}

func WriteEncryptedHeaderBytes(uuidStr string, plaintext, key []byte) ([]byte, error) {
	iv := key[:16]
	ciphertext, err := CFB8Encrypt(key, iv, plaintext)
	if err != nil {
		return nil, err
	}

	uuidBytes := []byte(uuidStr)
	header := make([]byte, HeaderSize)

	binary.LittleEndian.PutUint32(header[0:4], 0)
	binary.LittleEndian.PutUint32(header[4:8], Magic)
	binary.LittleEndian.PutUint64(header[8:16], 0)

	header[16] = byte(len(uuidBytes))
	copy(header[17:], uuidBytes)

	result := make([]byte, 0, HeaderSize+len(ciphertext))
	result = append(result, header...)
	result = append(result, ciphertext...)
	return result, nil
}

func CFB8DecryptData(data, fileKey []byte) ([]byte, error) {
	if len(data) >= 8 && binary.LittleEndian.Uint32(data[4:8]) == Magic {
		ciphertext := data[HeaderSize:]
		key := SkinpackKey
		if fileKey != nil {
			key = fileKey
		}
		return CFB8Decrypt(key, key[:16], ciphertext)
	}
	if fileKey != nil {
		return CFB8Decrypt(fileKey, fileKey[:16], data)
	}
	return nil, fmt.Errorf("no header and no key provided")
}
